crates/pallet-domains/src/bundle_storage_fund.rs

```diff
@@ -0,0 +1,219 @@
+    operator_id: OperatorId,
+    bundle_size: u32,
+) -> Result<(), Error> {
+    if bundle_size.is_zero() {
+        return Ok(());
+    }
+
+    let storage_fund_acc = storage_fund_account::<T>(operator_id);
+    let storage_fee = T::StorageFee::transaction_byte_fee() * bundle_size.into();
+
+    T::Currency::burn_from(
+        &storage_fund_acc,
+        storage_fee,
+        Precision::Exact,
+        Fortitude::Polite,
+    )
+    .map_err(|_| Error::BundleStorageFeePayment)?;
+
+    // Note the storage fee, it will go to the consensus block author
+    T::StorageFee::note_storage_fees(storage_fee);
+
+    Ok(())
+}
+
+/// Refund the front paid storage fee of a particular domain block back to the operator, the amount to
+/// refund to a particular operator is determined by the total storage fee collected from the domain user
+/// and the percentage of bundle storage that the operator have submitted for the domain block.
+#[allow(dead_code)]
+pub fn refund_storage_fee<T: Config>(
+    total_storage_fee: BalanceOf<T>,
+    paid_bundle_storage_fees: BTreeMap<OperatorId, u32>,
+) -> Result<(), Error> {
+    if total_storage_fee.is_zero() {
+        return Ok(());
+    }
+
+    let total_paid_storage = paid_bundle_storage_fees.values().sum::<u32>();
+    let mut remaining_fee = total_storage_fee;
+    for (operator_id, paid_storage) in paid_bundle_storage_fees {
+        // If the operator is deregistered and unlocked or slashed and finalized, the refund bundle storage
+        // fee will go to the treasury
+        if Operators::<T>::get(operator_id).is_none() || paid_storage.is_zero() {
+            continue;
+        }
+
+        let refund_amount = {
+            let paid_storage_percentage = Perbill::from_rational(paid_storage, total_paid_storage);
+            paid_storage_percentage.mul_floor(total_storage_fee)
+        };
+        let storage_fund_acc = storage_fund_account::<T>(operator_id);
+        T::Currency::mint_into(&storage_fund_acc, refund_amount).map_err(|_| Error::MintBalance)?;
+
+        remaining_fee = remaining_fee
+            .checked_sub(&refund_amount)
+            .ok_or(Error::BalanceUnderflow)?;
+    }
+
+    // Drop any dust and deregistered/slashed operator's bundle storage fee to the treasury
+    if !remaining_fee.is_zero() {
+        T::Currency::mint_into(&T::TreasuryAccount::get(), remaining_fee)
+            .map_err(|_| Error::MintBalance)?;
+    }
+
+    Ok(())
+}
+
+/// Split the new deposit into 2 parts: the staking deposit and the the storage fee deposit,
+/// add the storage fee deposit to the bundle storage fund.
+pub fn deposit_reserve_for_storage_fund<T: Config>(
+    operator_id: OperatorId,
+    source: &T::AccountId,
+    deposit_amount: BalanceOf<T>,
+) -> Result<NewDeposit<BalanceOf<T>>, Error> {
+    let storage_fund_acc = storage_fund_account::<T>(operator_id);
+
+    let storage_fee_reserve = STORAGE_FEE_RESERVE.mul_floor(deposit_amount);
+
+    T::Currency::transfer(
```

Comment:
We should also emit an event for this so that this amount can be indexed later and also gives some info to the user